HPC-9975 Remove `use-query-params`

Onitoxan opened the pull request:
## Nature of PR

- Bug-fix
- Hot-fix
- 🟢 Feature
- Testing
- 🟢 Package update
- CI


## Description

This PR is related to HPC-9965, read for more context.

In this PR we focus on removing `use-query-params` package. It became obsolete since with the use of `react-router` data API hooks (precisely, `useSearchParams()`) and using `io-ts` to write runtime validation, we can achieve the same objective.

We do this in order to reduce dependency of external libraries.

## How to test changes

To test changes I would try to manipulate the URL to see how it reacts even to unexpected `params` and with wrong data on it.

Beside this, I would focus mainly on reading logic for `useQueryParams()` to verify it really covers corner cases

Onitoxan committed:
Add custom `useQueryParams` hook

It's built on top of `useSearchParams` from `react-router`.

Onitoxan committed:
Add codecs for validation

Onitoxan committed:
Replace all usages of `use-query-params` for new `useQueryParams`

Onitoxan committed:
Remove `QueryParamProvider` and adapter

Onitoxan committed:
Uninstall `use-query-params` package

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/useQueryParams.tsx:
```diff
@@ -0,0 +1,69 @@
+
+    const result = codec.decode(obj);
+
+    if (!isRight(result)) {
+      console.error('Invalid query params:', result);
+      console.warn('Reverting back to initial values...');
+      return initialValues;
+    }
```
This made me thinking about what we should do if invalid parameters are detected in the URL. I don't think we should just log to console and let it be. The console is for developers and the question should be asked to clients if they want users to be informed of bad requests and how.

I'd say some violations are definitely worth throwing BadRequest error, like when string is used where number is expected, which already happens on `develop` with `rowsPerPage` query param, I think.

This is worth a bigger discussion with clients, because there are a lot of things that need to be defined.

Onitoxan replied:
I see, I have posed this question to client, once I have a response I will write back in this comment to see what is the resolution for this issue

Onitoxan replied:
After speaking with Kashif, taking into consideration clients view also, its fine how is working right now. Since its a private application, with very few users and no one will be modifying URLs anyways, its okay if it defaults to default values if anything missmatches

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/codecs.ts:
```diff
@@ -0,0 +1,65 @@
+    filters: t.string,
+    tableHeaders: t.string,
+  }),
+  t.partial({
+    prevPageCursor: util.INTEGER_FROM_STRING,
+    nextPageCursor: util.INTEGER_FROM_STRING,
+  }),
```
It should be noted that paging doesn't work without cursors, even though it appears by looking in the UI label saying something like "51–100 of 395432". This label is constructed via `page` query param, but API doesn't return proper results without cursors.

This problem is not created by this PR, but already exists on `develop`. We should ticket it separately.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/useQueryParams.tsx:
```diff
@@ -0,0 +1,69 @@
+    return result.right;
+  };
+
+  const setValidatedSearchParams = (
+    newParams: T,
+    flushSync?: NavigateOptions['flushSync']
+  ) => {
```
I am afraid of using this `flushSync` option. If you look at the [`react-router` PR which introduced this option](https://github.com/remix-run/react-router/pull/11005/files), you will see that it passes this option to React DOM and [docs about this option](https://react.dev/reference/react-dom/flushSync) are really trying to warn against its usage. The last sentence says:
> Most of the time, `flushSync` can be avoided, so use `flushSync` as a last resort.

Onitoxan replied:
I gave it another thought, and I found an alternative solution that requires the use of `useTimeou()`.

Long story short, resetting `Formik` form and setting query params are two actions that trigger a re-render, since both were being executed at the same time they created weird behaviors in React's loop. By adding `useTimeout()` without `delay` param (It defaults to 1ms) we ensure it gets executed in the next clock tic.

Another solution could be to add an `useEffect()` and add in the dependency array to trigger when the Formik form state changes. I find this approach more verbose, and more prone to bugs.

So I will push the solution using `useTimeout()`

Pl217 replied:
Just a note, you are writing about `useTimeout`, which sounds like a hook, which I didn't know of. But, you actually are using [`setTimeout()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/setTimeout) method, which doesn't default to 1, but when second parameter is omitted:
> ... a value of 0 is used, meaning execute "immediately", or more accurately, the next event cycle.

Onitoxan replied:
Its indeed setTimeout

Pl217 commented on apps/hpc-ftsadmin/src/app/pages/flows/flows-list.tsx:
```diff
@@ -51,44 +43,23 @@ export default (props: Props) => {
-    tableHeaders: withDefault(StringParam, encodeTableHeaders([])), // Default value of table headers
-    prevPageCursor: withDefault(NumberParam, undefined),
-    nextPageCursor: withDefault(NumberParam, undefined),
+    codec: FLOW_PARAMS_CODEC,
+    initialValues: {
+      page: 0,
+      rowsPerPage: 50,
+      orderBy: 'flow.updatedAt',
```
This block of changes is replacing decoding of `use-query-params` with our own `io-ts` decoding, using `FLOW_PARAMS_CODEC` from above. In this `initialValues` parameter, you are just setting the defaults, but the custom codecs used to do a little more before.

For `rowsPerPage`, it used to limit to value `100`, if you manually enter higher value in the URL. This query param can be tackled in wider discussion of manual edits in URL bar.

But, if we looked at the `orderBy` param, it used to only allow those which are sortable, which means all but `details`. If you use this value, it fails differently compared to using any random value like `abcd`. This problem should be tackled in this PR.

Onitoxan replied:
I decided for `rowsPerPage` to limit the correct options to `[10, 25, 50, 100]` instead of just limiting maximum number to 100. If using any other value than this, MUI table rowsPerPage component breaks.

As for orderBy I added to check if its sortable as you said :)

Onitoxan committed:
Remove `flushSync` option

The proposed solution is to use `setTimeout`, reseting the form
and changing the URL are two actions that trigger a re-render.
In order to avoid missbehaviours on how they get added to React's
event queue, we add a `setTimeout` so the execution is done in the
next tick.

Onitoxan committed:
Update codecs to have more restrictions

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove pageCursors

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan commented:
This PR is blocked by ticket HPC-10019, once the work in here is merged we can proceed on merging this PR. It can be reviewed in the meantime.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Add page to endpoint parameters

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-organization-table.tsx:
```diff
@@ -69,10 +69,12 @@ export const FilterOrganizationsTable = (props: Props) => {
-      ...query,
-      page: 0,
-      filters: encodeFilters({}, ORGANIZATIONS_FILTER_INITIAL_VALUES),
+    setTimeout(() => {
```
I think it's worth having a comment explaining the usage of `setTimeout` here and in other 2 places

Onitoxan replied:
I absolutely agree, you are right. What do you think of:

```typescript
  //  We need to delay this action in a synchronous way to avoid
  //  making 2 setState() actions in an uncontrolled way that could
  //  mess with internal React's component update cycle
```

I know you like to very specifically write things, so if you have a better proposal, I can change it

Pl217 replied:
It's good, but do you mean `setState()` or `setQuery()`, around which the `setTimeout` is wrapped?

Onitoxan replied:
I wanted to refer to setState actions as a whole, not the specific `setQuery()`, since before this we do `formikResetForm()` and after `setQuery()` I want to refer to this kind of calls to `setState()` functions that update components

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/codecs.ts:
```diff
@@ -0,0 +1,83 @@
+} from './table-headers';
+
+const ROWS_PER_PAGE_OPTIONS = [10, 25, 50, 100] as const;
+const ROWS_PER_PAGE = new t.Type<number, number>(
```
You may want to explore making this a generic codec, which would test whether a number is among those in a set of allowed integers. I was thinking about names too, maybe name it `ALLOWED_INTEGER`

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/codecs.ts:
```diff
@@ -0,0 +1,83 @@
+        ? t.success(v)
+        : t.failure(v, c);
+    } else if (typeof v === 'string') {
+      return /^[0-9]+$/.test(v) &&
```
You can substitute `[0-9]` for `\d`
```suggestion
      return /^\d+$/.test(v) &&
```

Pl217 reviewed (changes requested):
I noticed you said "try manipulating the URL" in the initial comment and when I did, you said in the inline comments that no one will change the URL manually :P

Onitoxan commented:
> I noticed you said "try manipulating the URL" in the initial comment and when I did, you said in the inline comments that no one will change the URL manually :P

I should have been more specific, no user will manipulate the URL, but nonetheless, us developers we should make sure that nothing too "broken" happens in any case :)

Onitoxan committed:
Add comments to `useTimeout()` usage

Onitoxan committed:
Refactor `ROWS_PER_PAGE` codec into a generic one

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
